pep-9999.rst

```diff
@@ -0,0 +1,294 @@
+               4 MAKE_FUNCTION            0
+               6 LOAD_FAST                0 (lst)
+               8 GET_ITER
+              10 CALL                     0
+              20 RETURN_VALUE
+
+   Disassembly of <code object <listcomp> at 0x...>:
+   2           0 RESUME                   0
+               2 BUILD_LIST               0
+               4 LOAD_FAST                0 (.0)
+         >>    6 FOR_ITER                 4 (to 18)
+              10 STORE_FAST               1 (x)
+              12 LOAD_FAST                1 (x)
+              14 LIST_APPEND              2
+              16 JUMP_BACKWARD            6 (to 6)
+         >>   18 END_FOR
+              20 RETURN_VALUE
+
+The bytecode for the comprehension is in a separate code object. Each time
+``f()`` is called, a new single-use function object is allocated (by
+``MAKE_FUNCTION``), called (allocating and then destroying a new frame on the
+Python stack), and then immediately thrown away.
+
+Under this PEP, the compiler will emit the following bytecode for ``f()``
+instead:
+
+.. code-block:: text
+
+  1           0 RESUME                   0
+
+  2           2 LOAD_FAST                0 (lst)
+              4 GET_ITER
+              6 LOAD_FAST_AND_CLEAR      1 (x)
+              8 SWAP                     2
+             10 BUILD_LIST               0
+             12 SWAP                     2
+        >>   14 FOR_ITER                 4 (to 26)
+             18 STORE_FAST               1 (x)
+             20 LOAD_FAST                1 (x)
+             22 LIST_APPEND              2
+             24 JUMP_BACKWARD            6 (to 14)
+        >>   26 END_FOR
+             28 SWAP                     2
+             30 STORE_FAST               1 (x)
+             32 RETURN_VALUE
+
+There is no longer a separate code object, nor creation of a single-use function
+object, nor any need to create and destroy a Python frame.
+
+Isolation of the ``x`` iteration variable is achieved by the combination of the
+new ``LOAD_FAST_AND_CLEAR`` opcode at offset ``6``, which saves any outer value
+of ``x`` on the stack before running the comprehension, and ``30 STORE_FAST``,
+which restores the outer value of ``x`` (if any) after running the
+comprehension.
+
+If the comprehension accesses variables from the outer scope, inlining avoids
+the need to place these variables in a cell, allowing the comprehension (and all
+other code in the outer function) to access them as normal fast locals instead.
+This provides further performance gains.
+
+This PEP only proposes to inline comprehensions occurring inside functions,
+where fast-locals (``LOAD_FAST/STORE_FAST``) are used. Module-level
+comprehensions will continue to create and call a function. Module-level
+comprehensions are generally called only once (when the module is imported), so
+optimizing the performance of calling them is unnecessary, and inlining them
+would require separate code paths in the compiler to handle a module global
```

Comment:
It would also have backward-compatibility implications:

```
x = 3

def f():
    return x

print([f() for x in (1, 2)])
```

If we temporarily clear `x` while the module-level comprehension is ongoing, this will produce `(1, 2)` instead of `(3, 3)`.